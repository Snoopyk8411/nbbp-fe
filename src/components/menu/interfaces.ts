import { Appearance } from 'tools/types/api-catalog-items-types';
import { SubmenuOpenActionType, SubmenuPosition } from './constants';

export interface IMenuProps {
  items: IMenuItemModel[];
  submenuPosition: SubmenuPosition;
}

export interface IMenuLevelProps {
  items: IMenuItemModel[];
  parentItem?: IMenuItemModel;
  submenuOpenActionType: SubmenuOpenActionType;
  submenuPosition: SubmenuPosition;
  containerElement: HTMLElement | null;
  onClose?: () => void;
}

export type MenuItemIdType = string | number;

export interface IMenuItemModel {
  id: MenuItemIdType;
  name: string;
  order: number;
  url: string;
  appearance: Appearance | String;
  children?: IMenuItemModel[];
  svgName: string | undefined;
}
