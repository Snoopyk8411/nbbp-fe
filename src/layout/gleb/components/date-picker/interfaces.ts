export interface IDateData {
  year: number;
  month: number;
  date: number;
}
