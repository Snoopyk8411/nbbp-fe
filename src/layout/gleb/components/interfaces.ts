import { AnyAction, Action, Store, Middleware } from 'redux';

export interface IMockStore<S = any, A extends Action = AnyAction> extends Store<S, A> {
  getActions(): any[];
  clearActions(): void;
}

export type MockStoreEnhanced<S = {}, DispatchExts = {}> = IMockStore<S> & { dispatch: DispatchExts };

export type MockStoreCreator<S = {}, DispatchExts = {}> = (
  state?: S | MockGetState<S>,
) => MockStoreEnhanced<S, DispatchExts>;

export type MockGetState<S = {}> = (actions: AnyAction[]) => S;

/**
 * Create Mock Store returns a function that will create a mock store from a state
 * with the same set of set of middleware applied.
 *
 * @param middlewares The list of middleware to be applied.
 * @template S The type of state to be held by the store.
 * @template DispatchExts The additional Dispatch signatures for the middlewares applied.
 */
declare function createMockStore<S, DispatchExts = {}>(middlewares?: Middleware[]): MockStoreCreator<S, DispatchExts>;

export default createMockStore;

export interface IMockData {
  mediaPage: {
    media: {
      copyright: string;
      date: string;
      explanation: string;
      hdurl: string;
      media_type: string;
      service_version: string;
      title: string;
      url: string;
    };
    error: Error | null;
    selectedDay: any;
    isLoading: boolean;
  };
}
