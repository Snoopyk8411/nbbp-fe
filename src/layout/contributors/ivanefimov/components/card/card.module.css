.card {
  --clr-neutral-900: hsl(207, 19%, 9%);
  --clr-neutral-200: hsl(0, 0%, 85%);
  --clr-neutral-100: hsl(0, 0%, 100%);
  --clr-accent-400: hsl(142, 90%, 61%);

  background-color: var(--clr-neutral-100);
  color: var(--clr-neutral-100);
  font-family: Tahoma, sans-serif;
  background-size: cover;
  padding: 160px 0 0;
  width: 240px;
  height: 360px;
  border-radius: 8px;
  overflow: hidden;

  transition: transform 0.4s ease;
}

.card:hover,
.card:focus-within {
  transform: scale(1.05);
}

.content {
  --padding: 16px;
  padding: var(--padding);
  padding-bottom: 24px;
  background: linear-gradient(hsl(0 0% 0% / 0), hsl(20 0% 0% / 0.3) 20%, hsl(0 0% 0% / 1));
}

.heading {
  position: relative;
  width: max-content;
}

.heading::after {
  content: '';
  position: absolute;
  height: 4px;
  background-color: var(--clr-accent-400);
  left: calc(var(--padding) * -1);
  bottom: -4px;
  width: calc(100% + var(--padding));

  transform-origin: left;
  transition: transform 0.4s ease;
}

.description {
  color: var(--clr-neutral-200);
}

.button {
  cursor: pointer;
  display: inline-block;
  text-decoration: none;
  text-transform: capitalize;
  font-weight: 700;
  color: var(--clr-neutral-900);
  background-color: var(--clr-accent-400);
  padding: 10px 25px;
  border-radius: 5px;
}

.button:active {
  transform: scale(0.95);
}

@media (hover) {
  .content {
    transform: translateY(60%);
    transition: transform 0.4s ease;
  }
  .card:hover .content,
  .card:focus-within .content {
    transform: translateY(0);
    transition-delay: 0.4s;
  }

  .card:focus-within .content {
    transition-duration: 0;
  }

  .content > *:not(.heading) {
    opacity: 0;
    transition: opacity 0.4s linear;
  }

  .card:hover .content > *:not(.heading),
  .card:focus-within .content > *:not(.heading) {
    opacity: 1;
    transition-delay: 0.65s;
  }

  .card:hover .heading::after,
  .card:focus-within .heading::after {
    transform: scaleX(1);
  }

  .heading::after {
    transform: scaleX(0);
  }

  .button:hover,
  .button.focus {
    background-color: var(--clr-neutral-100);
  }
}
