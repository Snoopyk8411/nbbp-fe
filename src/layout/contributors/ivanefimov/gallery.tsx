import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import useVisibility from 'hooks/use-visibility';
import { fetchPage } from 'store/contributors/ivanefimov/slice';
import { selectError, selectHasMore, selectIsLoading, selectPhotos } from 'store/contributors/ivanefimov/selectors';
import { TITLE } from 'store/contributors/ivanefimov/constants';
import { Card } from './components/card/card';
import { Loader } from './components/loader/loader';
import { IPhoto } from './interfaces';

import galleryStyles from './gallery.module.css';

interface IGalleryProps {
  initialPhotos: IPhoto[];
}

const Gallery = ({ initialPhotos }: IGalleryProps): JSX.Element => {
  const [isEndOfPage, endRef] = useVisibility<HTMLDivElement>();

  const dispatch = useDispatch();

  const photos = useSelector(selectPhotos);
  const isLoading = useSelector(selectIsLoading);
  const error = useSelector(selectError);
  const hasMore = useSelector(selectHasMore);

  const isFetchNextPageRequired = !isLoading && hasMore && isEndOfPage;

  useEffect(() => {
    if (isFetchNextPageRequired) {
      dispatch(fetchPage());
    }
  }, [isEndOfPage]);

  return (
    <>
      <div className={galleryStyles.container}>
        <h1 className={galleryStyles.heading}>{TITLE}</h1>
        {error && <p className={galleryStyles.error}>{error.message}</p>}
        <div className={galleryStyles.cards}>
          {initialPhotos.map(photo => (
            <Card key={photo.id} photo={photo} />
          ))}
          {photos.map(photo => (
            <Card key={photo.id} photo={photo} />
          ))}
          {isLoading && <Loader />}
        </div>
        <div ref={endRef}></div>
      </div>
    </>
  );
};

export default Gallery;
