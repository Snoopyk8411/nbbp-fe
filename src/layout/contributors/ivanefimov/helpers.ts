import { BASE_URL, CardSize } from '../../../store/contributors/ivanefimov/constants';

export const getUrlById = (id: string) => `url(${BASE_URL}/id/${id}/${CardSize.WIDTH}/${CardSize.HEIGHT})`;
