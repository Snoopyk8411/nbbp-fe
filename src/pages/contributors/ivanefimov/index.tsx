import Gallery from 'layout/contributors/ivanefimov/gallery';
import { IPhoto } from 'layout/contributors/ivanefimov/interfaces';
import { apiGetPhotos } from 'store/contributors/ivanefimov/api-get-photos';

interface IIndexPageProps {
  initialPhotos: IPhoto[];
}

const IndexPage = ({ initialPhotos }: IIndexPageProps): JSX.Element => {
  return <Gallery initialPhotos={initialPhotos} />;
};

export default IndexPage;

export const getServerSideProps = async () => {
  const initialPhotos = await apiGetPhotos();
  return {
    props: { initialPhotos },
  };
};
