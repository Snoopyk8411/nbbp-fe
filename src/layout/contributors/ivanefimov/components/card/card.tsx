import Link from 'next/link';
import { useRouter } from 'next/router';

import { IPhoto } from 'layout/contributors/ivanefimov/interfaces';
import { getUrlById } from 'layout/contributors/ivanefimov/helpers';
import cardStyles from './card.module.css';

interface ICardProps {
  photo: IPhoto;
}

export const Card = ({ photo }: ICardProps): JSX.Element => {
  const router = useRouter();
  return (
    <div className={cardStyles.card} style={{ backgroundImage: getUrlById(photo.id) }}>
      <div className={cardStyles.content}>
        <h3 className={cardStyles.heading}>by {photo.author}</h3>
        <p className={cardStyles.description}>Original size:</p>
        <p className={cardStyles.description}>
          {photo.width} x {photo.height} px
        </p>
        <Link href={`${router.asPath}/${photo.id}`}>
          <a className={cardStyles.button}>Download</a>
        </Link>
      </div>
    </div>
  );
};
