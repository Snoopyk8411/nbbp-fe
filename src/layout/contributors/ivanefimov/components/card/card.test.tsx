import renderer from 'react-test-renderer';

import { Card } from './card';
import { IPhoto } from 'layout/contributors/ivanefimov/interfaces';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const useRouter = jest.spyOn(require('next/router'), 'useRouter');
useRouter.mockImplementation(() => ({
  asPath: '',
}));

describe('Card component', () => {
  it('should get valid snapshot', () => {
    const photo: IPhoto = {
      id: '1',
      author: 'author',
      url: 'url',
      width: 100,
      height: 100,
      download_url: 'url',
    };
    const component = renderer.create(<Card photo={photo} />);
    expect(component).toMatchSnapshot();
  });
});
