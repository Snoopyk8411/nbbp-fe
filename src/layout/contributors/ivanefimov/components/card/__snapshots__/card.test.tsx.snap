// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should get valid snapshot 1`] = `
<div
  className="card"
  style={
    Object {
      "backgroundImage": "url(https://picsum.photos/id/1/200/300)",
    }
  }
>
  <div
    className="content"
  >
    <h3
      className="heading"
    >
      by 
      author
    </h3>
    <p
      className="description"
    >
      Original size:
    </p>
    <p
      className="description"
    >
      100
       x 
      100
       px
    </p>
    <a
      className="button"
      href="/1"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Download
    </a>
  </div>
</div>
`;
